packages/loader/container-loader/src/virtualDeltaManager.ts

```diff
@@ -0,0 +1,278 @@
+import { IDocumentService } from "@fluidframework/driver-definitions";
+import { IDocumentMessage, ISequencedDocumentMessage } from "@fluidframework/protocol-definitions";
+import { IConnectionManager, IConnectionManagerFactoryArgs } from "./contracts";
+import { DeltaManager } from "./deltaManager";
+
+const GroupedBatchOpType = "groupedBatch";
+
+class VirtualDeltaQueueProxy<T> extends TypedEventEmitter<IDeltaQueueEvents<T>> implements IDeltaQueue<T> {
+    private readonly _trackedListeners: Map<string, (...args: any[]) => void> = new Map();
+
+    constructor(
+        private readonly _queue: IDeltaQueue<T>,
+        private readonly _virtualizeValue: (value: T, storeValue?: boolean) => T,
+    ) {
+        super();
+
+        this.on("newListener", (event, _listener) => {
+            assert((this.listeners(event).length !== 0) === this._trackedListeners.has(event), "mismatch");
+            if (!this._trackedListeners.has(event)) {
+                let listener: (...args: any[]) => void;
+
+                // eslint-disable-next-line unicorn/prefer-ternary
+                if (event === "push" || event === "op") {
+                    listener = (task: T) => { this.emit(event, this._virtualizeValue(task, true /* storeValue */)); }
+                } else {
+                    listener = (...args: any[]) => { this.emit(event, ...args); };
+                }
+
+                this.addTrackedListener(event, listener);
+            }
+        });
+    }
+
+    private addTrackedListener(event: string, listener: (...args: any[]) => void) {
+        (this._queue as any).on(event, listener); // TODO: find better cast than "any"
+        assert(!this._trackedListeners.has(event), "double tracked listener");
+        this._trackedListeners.set(event, listener);
+    }
+
+    private removeTrackedListeners() {
+        for (const [event, listener] of this._trackedListeners.entries()) {
+            (this._queue as any).off(event, listener); // TODO: find better cast than "any"
+        }
+        this._trackedListeners.clear();
+    }
+
+    public peek(): T | undefined {
+        const value = this._queue.peek();
+        return value !== undefined ? this._virtualizeValue(value) : undefined;
+    }
+
+    public toArray(): T[] {
+        return this._queue.toArray().map((it) => this._virtualizeValue(it));
+    }
+
+    public dispose(error?: Error | undefined): void {
+        this.removeTrackedListeners();
+        return this._queue.dispose(error);
+    }
+
+    public get paused(): boolean {
+        return this._queue.paused;
+    }
+
+    public get length(): number {
+        return this._queue.length;
+    }
+
+    public get idle(): boolean {
+        return this._queue.idle;
+    }
+
+    public get disposed(): boolean {
+        return this._queue.disposed;
+    }
+
+    public async pause(): Promise<void> {
```

Comment:
Can you use DeltaQueueProxy as a base class here and avoid overwriting methods that are pass-through?

Reply:
I'm not sure I can since DeltaQueueProxy forwards events whereas I want to intercept them and virtualize before re-emitting

Reply:
Sorry, I meant DeltaManagerProxy.